fixed problem with padding samples up to integral number of frames

progaddict opened the pull request:
Hello!

I had a problem with padding while running the script `chime_data.py` on the CHiME3 dataset. The code supplied negative numbers to the `np.pad` function.

I believe, it is a bug because the function `_samples_to_stft_frames` cannot return more frames than necessary i.e. the restored number of samples, which is:

`samples = _stft_frames_to_samples(frames, size, shift)`

cannot exceed the current number of samples which is `time_signal.shape[time_dim]`

The new number of samples can be calculated by adding an additional frame:

`newSamples = _stft_frames_to_samples(frames + 1, size, shift)`

jheymann85 commented:
Hi,

thanks for the PR. But I don't see why `np.pad` should receive a negative number. We have

`frames = ceil((samples - size + shift) / shift)` which is >= `(samples - size + shift) / shift`.
so
`ceil((samples - size + shift) / shift) * shift + size - shift >= (samples - size + shift) / shift * shift + size - shift`
and thus
`ceil((samples - size + shift) / shift) * shift + size - shift >= samples = time_signal.shape[time_dim]`.

The problem might be that the code uses `numpy.ceil` which provides float values which might not be compatible with `np.pad` for newer versions of numpy. Could you check if changing `np.ceil` to `math.ceil` also solves the problem?

Thanks

progaddict commented:
I have added a test which just creates a dummy 6-channel noise and passes it to stft. The test fails for the original implementation (branch "tests"):

```
ERROR: testSyntheticNoise_StftShouldNotFail (__main__.TestStft)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/signal_processing/stft.py", line 11, in testSyntheticNoise_StftShouldNotFail
    spectrogram = stft(noise)
  File "/home/alex/py-projects/nn-gev/fgnt/signal_processing.py", line 138, in stft
    time_signal = np.pad(time_signal, pad, mode='constant')
  File "/home/alex/py-envs/py2scipy/local/lib/python2.7/site-packages/numpy/lib/arraypad.py", line 1318, in pad
    pad_width = _validate_lengths(narray, pad_width)
  File "/home/alex/py-envs/py2scipy/local/lib/python2.7/site-packages/numpy/lib/arraypad.py", line 1109, in _validate_lengths
    raise ValueError(fmt % (number_elements,))
ValueError: [(0, 0), [0, -98]] cannot contain negative values.
```

The same test passes with the new implementation (branch "master").

I have tested under python 2. However, it is interesting to note that **the same test did not fail** when I switched to python 3 (using again branch "tests" which contains the original implementation). So, as you have noted, it might be connected with implementation differences of `np.ceil` between python versions. I am going to test it as well.

progaddict commented:
I have received the same result for `math.ceil` (branch math-ceil): the test fails with the same error (`ValueError: [(0, 0), [0, -98]] cannot contain negative values.`) under python 2 and passes under python 3.

progaddict commented:
Oh, the test fails under python 3 for the corrected version! :anguished: 
OK, it seems it has something to do with the differences across python versions.
I think my correction belongs to the py2 branch so I'll move it there and close the PR.
But thanks for the hint about differences between possible implementations! :smiley: 
I haven' thought about that at the beginning.

jheymann85 commented:
Ah, I see, you are using Python 2.x. Then my guess is that it is a problem with the division rather than the veil operation since py3 is doing float division by default while py2 requires casting to float before the division.

progaddict commented:
Thanks a lot for the explanation!
